source/extensions/clusters/common/logical_host.h

```diff
@@ -61,6 +61,14 @@ class LogicalHost : public HostImpl {
     absl::ReaderMutexLock lock(&address_lock_);
     return HostImpl::address();
   }
+
+  const std::pair<Network::Address::InstanceConstSharedPtr,
+                  const std::vector<Network::Address::InstanceConstSharedPtr>>
+  addressAndListCopy() const {
+    absl::ReaderMutexLock lock(&address_lock_);
+    return {HostImpl::address(), HostImpl::addressList()};
```

Comment:
I think it's still hard to be sure this is safe. There's no protection from someone calling HostImpl::setAddressLiset concurrent with this locked access.

I know this is a lot less likely to crash because at the time you are reading this you have a lock, but that doesn't prevent a concurrent write, leading to undefined behavior. I know your test didn't catch it, and it might not be able to.

I don't know that it actually can happen now, but there's nothing stopping it from starting to happen in the future, and looking at upstream_impl.cc and .h there's no hint that this might be an issue.

One solution might be to virtualize the setters and hold the address locks inside them. But I think it would be nice if we could use absl thread annotations to be sure an error doesn't creep in later. In that case it might be better to keep separate member variables that use ABSL_GUARDED_BY in LogicalHost, and use the member vars in HostImpl only as a read source during init. WDYT?

Reply:
I agree that thread synchronization needs to be pushed into the base class, where the data member is defined.  It should not be an issue, since uncontended mutex lock is very cheap. But it will allow static annotations, preventing from future mistakes in this area.

Reply:
Currently there is no code directly call HostImpl::setAddressLiset() for logical host, other than the logical host c'tor, which is okay.  If in the future somebody add that in DNS refreshing scenario, then hopefully it will be caught by the added integration tests which do host refresh every 1ms.

Reply:
I would rather not rely on the integration test in that case. THis is what thread annotations are for.

If we decide not to pick one of the two options that let us use thread annotations, we should document heavily what the assumptions for, rename setAddressList to something very scary like setAddressListUnprotectedRace and put it in the `protected` section.

Reply:
@yanavlasov for your comments to add the mutex in the base class and protect its write/read, that will penalize the performance for all  other type of hosts, like static host as well, which IMHO should be avoided.

As far as I know, the current issue is Logical Host only as only Logical Host call setNewAddresses() after the host is created: https://github.com/envoyproxy/envoy/blob/5fc7662fb807b5ad0e882532857d102bb5c705df/source/extensions/clusters/logical_dns/logical_dns_cluster.cc#L142